filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsPathBuilder.scala

```diff
@@ -15,9 +15,24 @@ case class DrsPathBuilder(fileSystemProvider: DrsCloudNioFileSystemProvider) ext
 
   override def name: String = "DRS"
 
+  /**
+    * Unlike other cloud providers where directories are identified with a trailing slash at the end like `gs://bucket/dir/`,
+    * DRS has a concept of bundles for directories (not supported yet). This sanitizes the path by removing the trailing '/'(s)
+    * so that the CloudNioFileSystemProvider & CloudNioPath does not treat a DRS path ending with '/' as a directory, otherwise
+    * it's size is returned as 0
+    */
+  private def removeTrailingSlashes(path: String): String = {
+    if (path.length > (drsScheme.length + 3)) { //3: length of '://'
+      val pathArray = path.split(s"://")
+      val transformedPath = pathArray(1).replaceAll("[/]+$", "")
```

Comment:
I'm not 100% sure what this block is doing, and if it's consistent. For some inputs it ends up throwing exceptions.

- The `if` checks the `length`, but not `contains`
  `"foobarbaz".length > 6`
- The `split` splits the string
  `"foobarbaz".split("://") == Array("foobarbaz")`
- The `replaceAll` seems to just be stripping all final slashes
  `Array("foobarbaz").apply(1).replaceAll("[/]+$", "") == AIOOBE`

If the path really must contain the scheme, is that checked elsewhere? If that's _not a requirement_, which i'm guessing by the `else path` below, then how about

```scala
private def removeTrailingSlashes(path: String): String = path.replaceAll("[/]+$", "")
```

Reply:
> If the path really must contain the scheme, is that checked elsewhere?

The path should contain the scheme. It is checked [here](https://github.com/broadinstitute/cromwell/blob/4d631a3ba80e6fbf3807d5a4be2adcfde9bbe756/filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsPathBuilder.scala#L34) inside the `build` function.

> private def removeTrailingSlashes(path: String): String = path.replaceAll("[/]+$", "")

I previously did implement it, but the problem was that if the input is `dos://`, after passing the check that the path should start with `drsScheme://` (which is dos, above link), it will move onto replacing all trailing slashes (in this case the ones that separate the path from scheme) with empty string. Hence when it tries to parse it as a URI instead of throwing `Expected authority at index 6: dos://` not found exception it throws `Expected scheme-specific part at index 4: dos:` which is incorrect as the path does contain a valid scheme. Hence the `if else block`. The `if` part checks that there is some path after `dos://` string.

Reply:
suggestions that would make me feel better about it:

1) change the type to `Try` (or `IO` if you can stomach the ripple effect, probably too big...) since the input type is so broad
2) move it into a companion object and take drsScheme as an argument, makes testing easier as you don't have to mock the provider
3)  Add a couple tests that it does what you want and blows up on everything else

Reply:
The new method is also tested along with the `build` method in `DrsPathBuilderSpec`. It covers both good/bad test paths.

Reply:
purely TOL but a regex could make all requirements explicit and in one place